File: ruby/muse/data_structures/LinkedQueue.rb
module LinkedQueue
  class Node
    attr_accessor :data, :next

    def initialize(element)
      @data = element
      @next = nil
    end
  end

  class LinkedQueue
    def initialize
      @head = nil
      @tail = nil
      @length = 0
    end

    def size
      @length
    end

    def isEmpty
      @length.zero?
    end

    def peek
      raise "[PANIC - NoSuchElement]" if @length.zero?

      @head.data
    end

    def enqueue(element)
      node = Node.new(element)

      if @length.zero?
        @head = node
      else
        @tail.next = node
      end

      @tail = node

      @length += 1
    end

    def dequeue
      raise "[PANIC - NoSuchElement]" if @length.zero?

      target = @head

      if @length == 1
        @head = nil
        @tail = nil
      else
        @head = target.next
      end

      target.data = nil

      @length -= 1
    end
  end
end

File: python/muse/data_structures/linked_queue.py
class Node[T]:
    def __init__(self, element: T):
        self.data: T | None = element
        self.next: Node[T] | None = None


class LinkedQueue[T]:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def size(self) -> int:
        return self.length

    def is_empty(self) -> bool:
        return self.length == 0

    def peek(self) -> T:
        if self.length == 0:
            raise RuntimeError("[PANIC - NoSuchElement]")

        return self.head.data

    def enqueue(self, element: T) -> None:
        node = Node(element)

        if self.length == 0:
            self.head = node
        else:
            self.tail.next = node

        self.tail = node

        self.length += 1

    def dequeue(self) -> None:
        if self.length == 0:
            raise RuntimeError("[PANIC - NoSuchElement]")

        target = self.head

        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            self.head = target.next

        target.data = None

        self.length -= 1
